src/transformers/models/mask2former/modeling_mask2former.py

```diff
@@ -0,0 +1,3393 @@
+                last_hidden_state=encoder_output.last_hidden_state,
+                spatial_shapes=spatial_shapes,
+                level_start_index=level_start_index, 
+                hidden_states=encoder_output.hidden_states, 
+                attentions=encoder_output.attentions)
+
+        if not return_dict:
+            output = tuple(value for value in [last_hidden_state, spatial_shapes, level_start_index, encoder_states, attentions] if value is not None)
+
+        return output
+        
+
+# copied and adapted from original implementation, also practically equal to DetrSinePositionEmbedding
+# Copied from transformers.models.maskformer.modeling_maskformer.MaskFormerSinePositionEmbedding with MaskFormer->Mask2Former
+class Mask2FormerSinePositionEmbedding(nn.Module):
+    """
+    This is a more standard version of the position embedding, very similar to the one used by the Attention is all you
+    need paper, generalized to work on images.
+    """
+
+    def __init__(
+        self, num_pos_feats: int = 64, temperature: int = 10000, normalize: bool = False, scale: Optional[float] = None
+    ):
+        super().__init__()
+        if scale is not None and normalize is False:
+            raise ValueError("normalize should be True if scale is passed")
+        self.num_pos_feats = num_pos_feats
+        self.temperature = temperature
+        self.normalize = normalize
+        self.scale = 2 * torch.pi if scale is None else scale
+
+    def forward(self, x: Tensor, mask: Optional[Tensor] = None) -> Tensor:
+        if mask is None:
+            mask = torch.zeros((x.size(0), x.size(2), x.size(3)), device=x.device, dtype=torch.bool)
+        not_mask = ~mask
+        y_embed = not_mask.cumsum(1, dtype=torch.float32)
+        x_embed = not_mask.cumsum(2, dtype=torch.float32)
+        if self.normalize:
+            eps = 1e-6
+            y_embed = y_embed / (y_embed[:, -1:, :] + eps) * self.scale
+            x_embed = x_embed / (x_embed[:, :, -1:] + eps) * self.scale
+
+        dim_t = torch.arange(self.num_pos_feats, dtype=torch.float32, device=x.device)
+        dim_t = self.temperature ** (2 * torch.div(dim_t, 2, rounding_mode="floor") / self.num_pos_feats)
+
+        pos_x = x_embed[:, :, :, None] / dim_t
+        pos_y = y_embed[:, :, :, None] / dim_t
+        pos_x = torch.stack((pos_x[:, :, :, 0::2].sin(), pos_x[:, :, :, 1::2].cos()), dim=4).flatten(3)
+        pos_y = torch.stack((pos_y[:, :, :, 0::2].sin(), pos_y[:, :, :, 1::2].cos()), dim=4).flatten(3)
+        pos = torch.cat((pos_y, pos_x), dim=3).permute(0, 3, 1, 2)
+        return pos
+
+
+class Mask2FormerPixelDecoder(nn.Module):
+    def __init__(self,
+        config: DeformableDetrConfig,
+        in_features: Dict = None,
+        feature_size: int = 256, 
+        mask_feature_size: int = 256,
+        feature_channels: List[int] = None,
+        feature_strides: List[int] = None,
+        common_stride: int = None):
```

Comment:
> For example, if we need to add a significant number of parameters to `DeformableDetrConfig` to define the Pixel Decoder, it'd be better to create a stand alone PixelDecoderConfig class in the configuration file and use it to initialize a more structured `MultiScaleDeformAttnEncoderModule`. 

Forgive me if this is a silly question but had a small doubt related to this suggested change. Since there are only 3 new parameters that I want to add to the existing `DeformableDetrConfig` to define the Pixel Decoder config...can I do something like shown below code snippet while initializing `Mask2FormerConfig` class?

```
self.pixel_decoder_config = pixel_decoder_config # where pixel_decoder_config is `DeformableDetrConfig` 
self.pixel_decoder_config.feature_strides = feature_strides
self.pixel_decoder_config.encoder_in_features = encoder_in_features
self.pixel_decoder_config.common_stride = common_stride
```
And then I can pass `Mask2FormerConfig.pixel_decoder_config` to the Pixel Decoder 

Or is creating a separate class for `PixelDecoderConfig` the recommended approach in such a scenario?

Reply:
> * Rename `MultiScaleDeformAttnEncoderModule` to `Mask2FormerMSDAModel` to conform with the naming conventions.
> * Move the content of this class and forward function to `Mask2FormerMSDAModel` such that you can initialize the specific pixel decoder and return the outputs (all hidden_states / multi_scale features)
>   self.fpn = MaskFormerFPNModel(*args, feature_size=feature_size, **kwargs)

Sorry again have a silly doubt here. Just wanted to double check to be sure.
The current hierarchy of classes is like this for the pixel decoder:
Mask2FormerPixelLevelModule -> Mask2FormerPixelDecoder() -> MultiScaleDeformAttnEncoderModule() -> DeformableDetrEncoder()

If I understand correctly, you're suggesting to combine `Mask2FormerPixelDecoder` and `MultiScaleDeformAttnEncoderModule` to form `Mask2FormerMSDAModel` so then the updated structure of classes for pixel decoder would be something like this:
Mask2FormerPixelLevelModule() -> Mask2FormerMSDAModel() -> DeformableDetrEncoder()

My only concern in the above scenario is that then the forward call of `Mask2FormerMSDAModel` might become a little lengthy. Would that be alright?

Reply:
> Forgive me if this is a silly question but had a small doubt related to this suggested change. Since there are only 3 new parameters that I want to add to the existing `DeformableDetrConfig` to define the Pixel Decoder config...can I do something like shown below code snippet while initializing `Mask2FormerConfig` class?
> 
> ```
> self.pixel_decoder_config = pixel_decoder_config # where pixel_decoder_config is `DeformableDetrConfig` 
> self.pixel_decoder_config.feature_strides = feature_strides
> self.pixel_decoder_config.encoder_in_features = encoder_in_features
> self.pixel_decoder_config.common_stride = common_stride
> ```
> 
> And then I can pass `Mask2FormerConfig.pixel_decoder_config` to the Pixel Decoder

Yes, that is totally acceptable! We are just aiming for easy to understand and modularized code.

Reply:
> Does this clarify it?

Yes this clarifies it. Thanks!